content/en/docs/instrumentation/js/serverless.md

```diff
@@ -0,0 +1,166 @@
+---
+title: Serverless
+weight: 8
+---
+
+This guide will show you how to get started with tracing serverless functions using OpenTelemetry instrumentation libraries.
+
+## AWS Lambda
+
+Following the next steps will allow your AWS Lambda functions to use OpenTelemetry to send traces to the configured backend.
+
+### Dependencies
+
+Create an empty package.json:
+
+```sh
+npm init -y
+```
+
+Install dependencies used by the example.
+
+```sh
+npm install \
+  @opentelemetry/api \
+  @opentelemetry/auto-instrumentations-node \
+  @opentelemetry/exporter-trace-otlp-http \
+  @opentelemetry/instrumentation \
+  @opentelemetry/sdk-trace-base \
+  @opentelemetry/sdk-trace-node
+```
+
+### AWS Lambda wrapper code
+
+This file will contain all the OpenTelemetry logic, which will enable tracing.
+Please save the following code as `lambda-wrapper.js`.
+
+```javascript
+/* lambda-wrapper.js */
+
+const api = require("@opentelemetry/api");
+const { BatchSpanProcessor } = require("@opentelemetry/sdk-trace-base");
+const {
+  OTLPTraceExporter,
+} = require("@opentelemetry/exporter-trace-otlp-http");
+const { NodeTracerProvider } = require("@opentelemetry/sdk-trace-node");
+const { registerInstrumentations } = require("@opentelemetry/instrumentation");
+const {
+  getNodeAutoInstrumentations,
+} = require("@opentelemetry/auto-instrumentations-node");
+
+api.diag.setLogger(new api.DiagConsoleLogger(), api.DiagLogLevel.ALL);
+
+const provider = new NodeTracerProvider();
+const collectorOptions = {
+  url: "<backend_url>",
+};
+
+const spanProcessor = new BatchSpanProcessor(
+  new OTLPTraceExporter(collectorOptions)
+);
```

Comment:
Hmmm. I am not sure about this. My understanding is that batches can get lost on AWS Lambda. Don't know about other serverless environments. But I'm not an expert here either.

Reply:
Probably there needs to be a [force flush](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#forceflush) somewhere?

Reply:
I think that can also hurt performance in general. I'd be okay with getting this out there as a first cut but we can't have things documented that drop spans or significantly screw up performance.

Maybe @Aneurysm9 has some suggestions on what to document here?

Reply:
The BatchSpanProcessor does flush on [shutdown](https://github.com/open-telemetry/opentelemetry-js/blob/27c4f762948c9b8d693d5ce1c2df5aaa5b0e92ee/packages/opentelemetry-sdk-trace-base/src/export/BatchSpanProcessorBase.ts#L99), so it should send all traces before the lambda shutdown.
I'm not sure how big of a performance difference it is compared to a SimpleSpanProcessor for example, but the BatchSpanProcessor is also used in the [OpenTelemetry AWS Lambda Layer](https://github.com/open-telemetry/opentelemetry-lambda/blob/3971af6d6f47e9dc138f9d45448880e3ba8ae37f/nodejs/packages/layer/src/wrapper.ts#L101)

Reply:
It is generally recommended to use the `BatchSpanProcessor` everywhere but in testing where blocking export on span end can be important and tolerated.  With Lambda functions it is generally advised to call `ForceFlush` on the `(Trace|Meter)Provider` at the end of your function invocation to ensure that any telemetry is emitted before the runtime can be frozen.

There are [known issues](https://github.com/open-telemetry/opentelemetry-lambda/issues/224) with telemetry getting dropped by the Lambda layers, but they relate to extension lifecycle management issues in the collector extension layer and should not be affected by choice of span processor in the function space.

Reply:
Makes sense! Yeah, in this case ensuring there's a call at the end of the function invocation to forceFlush should then suffice.